### ISSUE TITLE
Type safe prepared statements in parser

### ISSUE THREAD

**Issue by brancz** [prometheus/prometheus#7197]
## Proposal 

Add syntax to the PromQL language, that allows making type safe prepared statements. I think syntax and actual functionality is heavily discussable so I wouldn't focus on this part just yet (I would rather first come to consensus that this type of feature is something we want rather than the nitty gritty details), but I imagine something along the lines of:

`rate(http_requests_total{%selector}[5m])`

Which reserves the templatable selector in the query's AST, and can be added at a later point. (potentially these could even be named variables, but that could also be a client concern and for the parser it's just a list of variables)

I haven't thought through possible clashes with other PromQL edge cases, obviously the syntax would need to be designed to not interfere with other PromQL features or break them in any way.

**Use case. Why is this important?**

The obvious use is prevention of typical query injection type attacks, but that's not the only use I see.

PromQL statements are being text-templated all over the place in the ecosystem. A list of examples (things that I know off the top of my head, and work with regularly so I feel the pain a lot):

* [monitoring mixins](https://github.com/monitoring-mixins/docs)
* [k8s-prometheus-adapter](https://github.com/DirectXMan12/k8s-prometheus-adapter/blob/4b5788e847e6b009d72b886f22d54f9363f9eba0/deploy/manifests/custom-metrics-config-map.yaml#L20)
* [argocd](https://argoproj.github.io/argo-rollouts/features/analysis/#analysis-at-a-predefined-step)

Alternatives that have been mentioned when discussing this on IRC was just doing text templating and seeing if the query parses. This is problematic, as in the above example uses if the surrounding PromQL query is incorrect it is only noticed at runtime, even though the failure could have been detected when writing the prepared PromQL statement. Both the query and input variables being unstructured also makes parser errors very difficult to understand as both query, input and or the combination of both could be the culprit of a malformed resulting query.

@slrtbtfs @brian-brazil @SuperQ @bwplotka @csmarchbanks @beorn7 @gouthamve @codesome @juliusv @tomwilkie @krasi-georgiev @simonpasquier

**Comment by juliusv** [comment]
I see the pain in the use case and agree it would be nice to be able to be able to copy-and-paste one of those mixin queries and check that it at least parses correctly with placeholders.

Just a thought: I wonder about the granularity level at which placeholders would be supported. It would seem easiest to make placeholders possible only as stand-ins for any current full PromQL AST nodes (of type string, scalar, vector, matrix). But your example for templating only part of a vector selector node would make things more complex. Then again, this seems like an extremely common use case in the mixins, so it would be good to support somehow.

**Comment by brian-brazil** [comment]
Is it proposed that such changes would be visible in the Prometheus binary, or just at the code level?

> Alternatives that have been mentioned when discussing this on IRC was just doing text templating and seeing if the query parses. This is problematic, as in the above example uses if the surrounding PromQL query is incorrect it is only noticed at runtime, even though the failure could have been detected when writing the prepared PromQL statement.

I'm confused by this claim. If you put in a dummy value you can see if the query parses in advance.


I'm not exactly seeing why a problem that appears solvable entirely within mixins given the existing parser needs fundamental language changes to the user-exposed PromQL.

**Comment by brancz** [comment]
> I'm confused by this claim. If you put in a dummy value you can see if the query parses in advance.

Parser errors would still be confusing, and it doesn't solve the injection case.

> Is it proposed that such changes would be visible in the Prometheus binary, or just at the code level?

I can see uses in Prometheus itself as well, but for a start I think just having it in the parser to allow ecosystem integrations would be sufficient to get some experience with it.

> I'm not exactly seeing why a problem that appears solvable entirely within mixins given the existing parser needs fundamental language changes to the user-exposed PromQL.

I don't think just because something is difficult or a lot of work, diminishes the value of a feature. Obviously someone must be willing to put in that work (should it be agreed on) and not break any existing features. Also I'd like to see how difficult this would actually be, before claiming something like this. My hope is that with the recent work from @slrtbtfs, this shouldn't be all that wild (I may be completely off though).

> But your example for templating only part of a vector selector node would make things more complex. Then again, this seems like an extremely common use case in the mixins, so it would be good to support somehow.

I honestly haven't looked too deep into implementation details yet, I generally agree though. I think we could even just start with what I feel are the two most common things:

* Selectors
* Groupings

**Comment by brian-brazil** [comment]
> Parser errors would still be confusing, 

I'm not following you here, you'd have the usual PromQL errors of the usual standard. Can you provide a concrete example?

> it doesn't solve the injection case.

I'm not following you here either, applying string escaping is not difficult.

> ecosystem integrations

Hmm, do any the mixins use the parser today? How does jsonnet integrate with Go code?

> I don't think just because something is difficult or a lot of work, diminishes the value of a feature.

I never said anything about the level of effort. My point was more why should problems external to Prometheus complicate Prometheus itself, as if the solution to every problem was to change Prometheus itself that's not likely to result in a coherent system.

**Comment by slrtbtfs** [comment]
Regarding the implementation difficulty:

The canonical way to achieve this probably would be to make individual selectors and groupings AST Nodes, too. This was already suggested in #6942 but not deemed necessary at that point in time. Since there would be some nontrivial refactoring across package boundaries and the addition of new node types involved, this probably would be the hardest part.

With that in place the parser could  be extended with a new node type that serves as a placeholder for things that would be inserted into queries. Since AFAIK PromQL doesn't have too much function operator type information can be added to those nodes, that can be used for type verification.

> I'm not following you here, you'd have the usual PromQL errors of the usual standard. Can you provide a concrete example?

When having a query like `foo{a=%b}` and claiming `%b` would be a label matcher (i.e. of the form `label="value",...`, what you're suggesting would be parsing something like `foo{a=label="value"}`. The resulting error message isn't really helpful: `1:7: parse error: unexpected identifier "label" in label matching, expected string`.

> I'm not following you here either, applying string escaping is not difficult.

When inserting more complex things like whole vector selectors, string escaping alone won't do it.

> I never said anything about the level of effort. My point was more why should problems external to Prometheus complicate Prometheus itself, as if the solution to every problem was to change Prometheus itself that's not likely to result in a coherent system.

The debate whether PromQL should be extended by something that can be already done with a template system seems to come up in a new Proposal every month or so and it seems to be more or less the same debate every time. Maybe some more general discussion about whether such features are desirable would be beneficial.

**Comment by brian-brazil** [comment]
> The resulting error message isn't really helpful: 

There's two problems here. One is verifying that `foo{a=%b}` is correct in itself, the other is that `%b` has an appropriate value.

Tracing the conversation this point is about the former, and substituting in `"param_b"` to check if the query parses doesn't seem difficult to me and will have the usual error message.

For the latter, the existing matcher parser can check that it's valid, and will have an appropriate error message.

I think a step has been missed in your explanations, is automatic type inference what you're going for or is it all statically typed by the user?

> Maybe some more general discussion about whether such features are desirable would be beneficial.

The question isn't whether some people desire them. The question is more what layer should they go in, and in particular if a given feature is appropriate for inclusion in Prometheus itself as a 1st class feature or should be built on top of it.

**Comment by slrtbtfs** [comment]
> I think a step has been missed in your explanations, is automatic type inference what you're going for or is it all statically typed by the user?

Both should be possible to implement with similar effort, but it doesn't really make sense to discuss about these details yet.

**Comment by brian-brazil** [comment]
It makes a big difference, as if you're going with user provided static types then what you want should already be doable with the parser today. Inferring the types from the query on the other hand I would imagine would require some pretty invasive changes.

**Comment by slrtbtfs** [comment]
If we extend the parser to parse place holders as such and do static type checking based on the types the user assigned to these place holders, the step to automatic type inference isn't that big any more.

But getting there in the first place would indeed be quite invasive.

**Comment by brian-brazil** [comment]
My point above is that you can do static type checking without having to change the parser at all. For a label value (i.e. string) try substituting in a quoted label value to the query and see if it parses. In the other direction try parsing it as a label value.

**Comment by juliusv** [comment]
Isn't the main use case just to be able to directly copy `rate(http_requests_total{%selector}[5m])` from one of the mixin files in your code editor and paste it into a PromQL input field that responds with "Yup this parses, but it's incomplete and to evaluate it you'll also need to provide the placeholder values"? Otherwise the editing flow is painful, as you'll have to manually replace all the selectors with something that yields a fully valid PromQL expression.

**Comment by beorn7** [comment]
Collected thoughts:

I have seen (and used) templating of many different parts of a PromQL expression. Some were, in a way, utilizing "injection attacks" by requiring fragments of PromQL to be inserted. Or in other words: In some use cases, the user doesn't even want type safety. (One might argue, that's bad usage, but that gives the discussion a different angle, from "we want to help users" to "we want to change bad behavior of users".)

Even if we assume type safety is desired, then my concern would be about the many many different types possible. A user needed to understand parsing and the AST first to even start using the feature effectively. Then there needed to be some kind of sub-language to specify the type. And even then, after all the efforts put into it by us and the user, some users might be dissatisfied, with being restricted to node types in the AST rather than saying: "This needs to be `bottom_k` or `top_k` with a 1st parameter less than 10`. Going the other way and don't let the user specify types at all (I guess that's meant by automatic inference above), we might not gain so much because a non-trivial template is in many cases expandable in a valid way by a non-intended type. I guess it boils down to a powerful typesafe templating system being a very deep rabbit hole…

I'm not quite sure if the prime goal here is checking un-expanded templates ("you want this to be expanded to a range selector, but that doesn't result in valid PromQL") or checking the parameters used for expansion ("this parameter is not a range selector") or both. In any case, this seems to be mostly about meaningful error messages. As already said above, how does that integrate with the templating engine used by the user? To deep-integrate with the editing workflow (I edit, for example, Jsonnet and get error messages as described above in my editor), I guess you needed a specific glue tool that understands both (for example) Jsonnet _and_ PromQL (syntax and AST types). Such a solution would be specific to a single templating engine, but as we see, users use many different templating engines. One step away from it would be what Julius hinted at: copy&pastie an unexpanded PromQL expression into a PromQL input field. But then you still have the problem of different templating syntax from different engines. And that's just the first part (about the un-expanded template), not the second about checking parameters. To work in a generic way with all templating engines, we could come up with our own intermediate templating syntax, which is then the one the user has to create with their templating engine (which resembles SQL prepared statements a bit, see the discussion about this being part of the SQL syntax, but note that SQL prepared statements are just about binding values while we, quite often, want to template PromQL _code_). But then you are still far away from an integrated editing experience, despite having inflicted the cost of another layer of templating…

Over all, I'm very sceptic about the "return of investment" here. However we turn this, it seems to be a lot of effort, the benefits of which will be hard to make generally usable.

**Comment by brian-brazil** [comment]
> (I guess that's meant by automatic inference above),

Yes, that was my meaning.

**Comment by juliusv** [comment]
> To work in a generic way with all templating engines, we could come up with our own intermediate templating syntax, which is then the one the user has to create with their templating engine

Just to clarify, my assumption was that the intention was indeed for the PromQL parser to have its own templating syntax (like the `rate(http_requests_total{%selector}[5m])` example), but in your config templating engine you would just write that entire expression out as a constant string and then provide the placeholder values out of band to the query evaluator. For example, in rule definitions this could be another section in parallel to the expression field:

```yaml
- alert: TheRateIsTooDamnHigh
  expr: rate(http_requests_total{%selector}[5m]) > 10
  placeholder_values:
    selector: 'job="api-server"'
```

And similar for the HTTP API, where the placeholder values could be passed as separate parameters.

> Over all, I'm very sceptic about the "return of investment" here. However we turn this, it seems to be a lot of effort, the benefits of which will be hard to make generally usable.

Yeah, it's starting to feel like that to me as well. On one hand templating selector matchers would be the most important use case, on the other hand just doing that feels "oddly specific" and partial on the implementation side. And allowing placeholders in all arbitrary places would be a lot of complexity.

**Comment by brancz** [comment]
Sounds like there is rough consensus that this is too high in complexity over the gain.
